sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiFlushAndFinalizeDoFn.java

```diff
@@ -170,6 +214,29 @@ public void process(PipelineOptions pipelineOptions, @Element KV<String, Operati
             BigQuerySinkMetrics.reportFailedRPCMetrics(
                 failedContext, BigQuerySinkMetrics.RpcMethod.FLUSH_ROWS);
 
+            if (error != null && isOffsetBeyondEndOfStreamError(error)) {
+              flushOperationsOffsetBeyondEnd.inc();
+              LOG.warn(
+                  "Flush of stream {} to offset {} failed because the offset is beyond the end of the stream. "
+                      + "This typically means the stream was finalized or truncated by BQ. "
+                      + "The operation will not be retried on this stream. Error: {}",
+                  streamId,
+                  offset,
+                  error.toString());
+              // This specific error is not retriable on the same stream.
+              // Throwing a runtime exception to break out of the RetryManager and signal
+              // to the Beam runner that the bundle should be retried, which will then
+              // allow an upstream DoFn to create a new stream.
+              throw new RuntimeException(
```

Comment:
If I understood correctly it does not fix the error, but processing and re-throw an clearer error. It still rely on runner retry. Is that correct?

Reply:
yes, it fails the entire bundle and track the error counts. I hope this could clean up the wrong stream. Maybe we could just `RetryType.DONT_RETRY`?

Reply:
Does the error mean that the data in this stream has already been finalized to BQ? Could there be any data missing or duplicated?